Feature: improve logging related to finding files

alhirzel opened the pull request:
## Description

Improves debug and trace logging when trying to locate files using the file resolver.

Showing the result of this change, note the basic output is unchanged:

```
alhirzel@gecko:/tmp $ mitsuba file_error.xml
2022-09-28 07:29:05 INFO  main  [mitsuba.cpp:340] Mitsuba version 3.0.2 (makepkg[d456826e], Linux, 64bit, 24 threads, 8-wide SIMD)
2022-09-28 07:29:05 INFO  main  [mitsuba.cpp:341] Copyright 2022, Realistic Graphics Lab, EPFL
2022-09-28 07:29:05 INFO  main  [mitsuba.cpp:342] Enabled processor features: cuda llvm avx2 avx fma f16c sse4.2 x86_64
2022-09-28 07:29:05 INFO  main  [xml.cpp:1405] Loading XML file "file_error.xml" with variant "scalar_rgb"..

Caught a critical exception: [xml.cpp:1112] Error while loading "file_error.xml" (near line 2, col 6): could not instantiate shape plugin of type "obj": 
[xml.cpp:1112]   [OBJMesh] Error while loading OBJ file "DOES_NOT_EXIST.obj": file not found
```

With `Debug` each lookup is annotated:

```
alhirzel@gecko:/tmp $ mitsuba -v file_error.xml
2022-09-28 07:29:09 INFO  main  [mitsuba.cpp:340] Mitsuba version 3.0.2 (makepkg[d456826e], Linux, 64bit, 24 threads, 8-wide SIMD)
2022-09-28 07:29:09 INFO  main  [mitsuba.cpp:341] Copyright 2022, Realistic Graphics Lab, EPFL
2022-09-28 07:29:09 INFO  main  [mitsuba.cpp:342] Enabled processor features: cuda llvm avx2 avx fma f16c sse4.2 x86_64
2022-09-28 07:29:09 INFO  main  [xml.cpp:1405] Loading XML file "file_error.xml" with variant "scalar_rgb"..
2022-09-28 07:29:09 DEBUG wrk0  [FileResolver] Looking for file "plugins/obj.so" on the resource search path(s)
2022-09-28 07:29:09 DEBUG wrk0  [PluginManager] Loading plugin "plugins/obj.so" ..
2022-09-28 07:29:09 DEBUG wrk0  [FileResolver] Looking for file "plugins/diffuse.so" on the resource search path(s)
2022-09-28 07:29:09 DEBUG wrk0  [PluginManager] Loading plugin "plugins/diffuse.so" ..
2022-09-28 07:29:09 DEBUG wrk0  [FileResolver] Looking for file "plugins/uniform.so" on the resource search path(s)
2022-09-28 07:29:09 DEBUG wrk0  [PluginManager] Loading plugin "plugins/uniform.so" ..
2022-09-28 07:29:09 DEBUG wrk0  [FileResolver] Looking for file "DOES_NOT_EXIST.obj" on the resource search path(s)
2022-09-28 07:29:09 DEBUG wrk0  [OBJMesh] Loading mesh from "DOES_NOT_EXIST.obj" ..

Caught a critical exception: [xml.cpp:1112] Error while loading "file_error.xml" (near line 2, col 6): could not instantiate shape plugin of type "obj": 
[xml.cpp:1112]   [OBJMesh] Error while loading OBJ file "DOES_NOT_EXIST.obj": file not found
```

And finally for `Trace` detailed info about where is checked:

```
alhirzel@gecko:/tmp $ mitsuba -vv file_error.xml
2022-09-28 07:29:13 INFO  main  [mitsuba.cpp:340] Mitsuba version 3.0.2 (makepkg[d456826e], Linux, 64bit, 24 threads, 8-wide SIMD)
2022-09-28 07:29:13 INFO  main  [mitsuba.cpp:341] Copyright 2022, Realistic Graphics Lab, EPFL
2022-09-28 07:29:13 INFO  main  [mitsuba.cpp:342] Enabled processor features: cuda llvm avx2 avx fma f16c sse4.2 x86_64
2022-09-28 07:29:13 INFO  main  [xml.cpp:1405] Loading XML file "file_error.xml" with variant "scalar_rgb"..
2022-09-28 07:29:13 DEBUG wrk0  [FileResolver] Looking for file "plugins/obj.so" on the resource search path(s)
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "plugins/obj.so" at /tmp/plugins/obj.so
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Found file "plugins/obj.so" at /usr/lib/mitsuba3-git/plugins/obj.so
2022-09-28 07:29:13 DEBUG wrk0  [PluginManager] Loading plugin "plugins/obj.so" ..
2022-09-28 07:29:13 DEBUG wrk0  [FileResolver] Looking for file "plugins/diffuse.so" on the resource search path(s)
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "plugins/diffuse.so" at /tmp/plugins/diffuse.so
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Found file "plugins/diffuse.so" at /usr/lib/mitsuba3-git/plugins/diffuse.so
2022-09-28 07:29:13 DEBUG wrk0  [PluginManager] Loading plugin "plugins/diffuse.so" ..
2022-09-28 07:29:13 DEBUG wrk0  [FileResolver] Looking for file "plugins/uniform.so" on the resource search path(s)
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "plugins/uniform.so" at /tmp/plugins/uniform.so
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Found file "plugins/uniform.so" at /usr/lib/mitsuba3-git/plugins/uniform.so
2022-09-28 07:29:13 DEBUG wrk0  [PluginManager] Loading plugin "plugins/uniform.so" ..
2022-09-28 07:29:13 DEBUG wrk0  [FileResolver] Looking for file "DOES_NOT_EXIST.obj" on the resource search path(s)
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "DOES_NOT_EXIST.obj" at /tmp/DOES_NOT_EXIST.obj
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "DOES_NOT_EXIST.obj" at /usr/lib/mitsuba3-git/DOES_NOT_EXIST.obj
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "DOES_NOT_EXIST.obj" at DOES_NOT_EXIST.obj
2022-09-28 07:29:13 TRACE wrk0  [FileResolver] Didn't find file "DOES_NOT_EXIST.obj" at DOES_NOT_EXIST.obj
2022-09-28 07:29:13 DEBUG wrk0  [OBJMesh] Loading mesh from "DOES_NOT_EXIST.obj" ..

Caught a critical exception: [xml.cpp:1112] Error while loading "file_error.xml" (near line 2, col 6): could not instantiate shape plugin of type "obj": 
[xml.cpp:1112]   [OBJMesh] Error while loading OBJ file "DOES_NOT_EXIST.obj": file not found
```

## Testing

I did not run the test suite because I am naively assuming this is a non-breaking change.

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `cuda_*` and `llvm_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

alhirzel committed:
improve logging related to finding files

Speierers commented on src/core/fresolver.cpp:
```diff
@@ -21,10 +22,15 @@ bool FileResolver::contains(const fs::path &p) const {
 
 fs::path FileResolver::resolve(const fs::path &path) const {
     if (!path.is_absolute()) {
+        Log(Debug, "Looking for file \"%s\" on the resource search path(s)", path.native());
```
Shouldn't this be logged into `Trace` as well?

alhirzel replied:
Currently `Debug` is written once per file lookup, and `Trace` is written once per path check. You're right that I forgot to paste the output--maybe it makes the behavior more clear.

Speierers replied:
IMO this is a little too much for debug mode. I would like to be able to compile in debug mode while not always having those logs. Let's use `Trace` here to have control. In any case, the log messages you are adding in this PR are useful to debug a situation where a file couldn't be found, in which case you will need to switch to `Trace` to see all the messages.

alhirzel replied:
Ok, agree, I will make all of the outputs part of `Trace`.

Speierers reviewed (commented):
Hi @alhirzel, thanks for this PR. 

Reading the PR description it seems that the output is missing.

alhirzel committed:
more logging improvements and guidance in docs

alhirzel commented:
I updated this to include a few more improvements that are in the following spirit (which I also added to the dev docs):

> When writing error or log messages that refer to an existing file, includ the full path to the file by default.

Speierers commented on src/core/plugin.cpp:
```diff
@@ -113,7 +113,7 @@ struct PluginManager::PluginManagerPrivate {
 
         if (fs::exists(resolved)) {
-            Log(Debug, "Loading plugin \"%s\" ..", filename.string());
+            Log(Debug, "Loading plugin \"%s\" from \"%s\" ..", name, resolved.string());
```
I don't think that's necessary.

alhirzel replied:
Ok - I plan to change it to use the full `resolved` path then. See what you think.

Speierers commented on src/core/spectrum.cpp:
```diff
@@ -13,9 +13,9 @@ void spectrum_from_file(const fs::path &path, std::vector<Scalar> &wavelengths,
     if (!fs::exists(file_path))
-        Log(Error, "\"%s\": file does not exist!", file_path);
+        Log(Error, "\"%s\": file does not exist!", file_path.string());
 
-    Log(Info, "Loading spectral data file \"%s\" ..", file_path);
+    Log(Info, "Loading spectral data file \"%s\" ..", file_path.string());
```
Does this fix a specific bug? What is the log message before and after this change?

alhirzel replied:
I actually made these types of changes during a very pedantic file path review for consistency in explicitly calling `.string()`; I don't know that it fixes a bug. I will back out this for less noise in the PR!

Speierers commented on src/emitters/envmap.cpp:
```diff
@@ -108,7 +108,7 @@ class EnvironmentMapEmitter final : public Emitter<Float, Spectrum> {
         if (bitmap->width() < 2 || bitmap->height() < 3)
             Throw("\"%s\": the environment map resolution must be at "
-                  "least 2x3 pixels", m_filename);
+                  "least 2x3 pixels", file_path.string());
```
Same question here.
